FEATURE/ElvaRivle/DfWCIPVU-v2

ElvaRivle opened the pull request:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * implemented login (on successful login JWT is returned)
 * added login service, controller and routes
--------------------------------------------------------------------
 OTHER
 * README typo
 * new JWT_SECRET env variable
 * closed mongoose connection on seed failure
 * globally enabled JSON body parsing
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * successfully connected to MongoDB
 * created User model
 * seeded default users to DB
--------------------------------------------------------------------
 OTHER
 * added all necessary packages
 * loading connection string from .env file
 * listening and logging mongoose events
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * seeder converted to script
 * README added
--------------------------------------------------------------------
 OTHER
 * changed default users for seeding
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * implemented login (on successful login JWT is returned)
 * added login service, controller and routes
--------------------------------------------------------------------
 OTHER
 * README typo
 * new JWT_SECRET env variable
 * closed mongoose connection on seed failure
 * globally enabled JSON body parsing
--------------------------------------------------------------------

ElvaRivle committed:
Merge branch 'develop' into FEATURE/ElvaRivle/DfWCIPVU-v2

becarselman commented on controllers/login.js:
```diff
@@ -0,0 +1,51 @@
+const loginUser = async (req, res) => {
+    const { email, password } = { ...req.body }
+
+    if (email === undefined) {
```
you can use !email and that will cover this case

becarselman commented on controllers/login.js:
```diff
@@ -0,0 +1,51 @@
+
+        if (user === null) {
+            return res.status(404).json({
+                error: "Email not found"
```
Can we found better message. I will suggest to create utils file with all error messages for example "Not found" and than you can reuse in other places.

becarselman reviewed (commented):
Just remove gitKeep files if they no longer needed

becarselman commented on controllers/login.js:
```diff
@@ -0,0 +1,51 @@
+const loginService = require("../services/login")
+
+
+const loginUser = async (req, res) => {
```
we discuss this already on meeting but I want to set as reminder. 

The "models" folder contains the data models, which define the structure and behavior of the data entities in the application.
The "repositories" folder contains the code for interacting with the database, such as CRUD (Create, Read, Update, Delete) operations, and database queries.
The "services" folder contains the business logic of the application. This includes any application-specific rules or algorithms that are used to process data or implement functionality.
The "controllers" folder contains the logic for handling HTTP requests and responses. Controllers receive requests from the client, process them using the services, and return a response to the client.
Finally, the "routes" folder contains the API endpoints and associated HTTP methods. Each route is mapped to a specific controller method, and is responsible for handling a specific request from the client.

becarselman commented on services/login.js:
```diff
@@ -0,0 +1,28 @@
+}
+
+function generateJWTToken(user) {
+    return jwt.sign({ email: user.email }, process.env.JWT_SECRET, { expiresIn: "1d" })
```
create new file where you can read all env vars and than import where you need

ElvaRivle committed:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * all errors added to errors.js file
 * all env variables added to env.js file
 * added user repository
--------------------------------------------------------------------
 OTHER
 * rewrote login service and controller
 * controller made minimalistic, all logic moved to service
--------------------------------------------------------------------

ElvaRivle commented:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * all errors added to errors.js file
 * all env variables added to env.js file
 * added user repository
--------------------------------------------------------------------
 OTHER
 * rewrote login service and controller
 * controller made minimalistic, all logic moved to service
--------------------------------------------------------------------

becarselman commented on controllers/login.js:
```diff
@@ -0,0 +1,25 @@
+
+        const { userId, token } = await loginService.loginUser(email, password)
+
+        return res.json({
```
I think you are missing status code here.

ElvaRivle replied:
It defaults to 200, but I will explicitly write it

becarselman commented on utils/env.js:
Find some different name for this. My suggestion is configuration

becarselman commented on db/models/User.js:
```diff
@@ -1,22 +1,20 @@
+        type: String,
+        required: true,
+        unique: true,
+        validate: [validator.isEmail, "Email in invalid format"]
```
You can reuse your new error messages in error file. Otherwise add new message : D

ElvaRivle replied:
yikes, forgot that one 😄

ElvaRivle committed:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * renamed `utils` folder to `configuration`
 * added EMAIL_INVALID_FORMAT error message
--------------------------------------------------------------------
 OTHER
 * explicitly returned 200 on successful login
 * fixed misc warnings about unused requires
--------------------------------------------------------------------

ElvaRivle commented:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
MAIN
* renamed `utils` folder to `configuration`
* added EMAIL_INVALID_FORMAT error message
--------------------------------------------------------------------
OTHER
* explicitly returned 200 on successful login
* fixed misc warnings about unused requires
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * when `nodemon` restarts, release previous connection back to pool
--------------------------------------------------------------------
 OTHER
 * capture SIGTERM as well
 * extracted app termination login in separate function

ElvaRivle commented:
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * when `nodemon` restarts, release previous connection back to pool
--------------------------------------------------------------------
 OTHER
 * capture SIGTERM as well
 * extracted app termination login in separate function
 * handling CTRL + C behaviour on Windows via `npm run` is a no-go, [reason](https://stackoverflow.com/a/74751308)
